server/src/daemon/secret_exchange.rs

```diff
@@ -20,55 +23,52 @@ pub struct SecretExchange {
 }
 
 impl SecretExchange {
-    pub fn secret_exchange_begin(&self) -> HashMap<&str, String> {
-        let mut exchange: HashMap<&str, String> = HashMap::new();
-        exchange.insert(PUBLIC, BASE64_STANDARD.encode(&self.public_key));
+    // Creates the initial payload containing caller public_key
+    pub fn secret_exchange_begin(&self) -> String {
+        let map = HashMap::from([(PUBLIC, self.public_key.as_ref())]);
 
-        exchange
+        encode(&map)
     }
 
-    pub fn secret_exchange_receive(
-        &self,
-        exchange_in: HashMap<&str, String>,
-    ) -> Result<HashMap<&str, String>, base64::DecodeError> {
-        let mut exchange_out: HashMap<&str, String> = HashMap::new();
-        let secret = exchange_in.get(SECRET);
-        if !secret.is_none() {
-            let secret = BASE64_STANDARD.decode(secret.unwrap())?;
-            let public_key = BASE64_STANDARD.decode(exchange_in.get(PUBLIC).unwrap())?;
-            let public_key = Key::new(public_key);
-            let iv = BASE64_STANDARD.decode(exchange_in.get(IV).unwrap())?;
+    // Decrypt and retrieve secret
+    pub fn secret_exchange_receive(&self, exchange: String) -> String {
+        let decoded = decode(&exchange).unwrap();
+        let mut encrypted: Vec<u8> = Vec::new();
+        let mut map: HashMap<&str, &[u8]> = HashMap::new();
 
-            let aes_key = Key::generate_aes_key(&self.private_key, &public_key);
-            let secret = crypto::decrypt(secret, &aes_key, &iv).to_vec();
+        let secret = decoded.get(SECRET);
+        if secret.is_some() {
+            let secret = secret.unwrap();
+            let public_key = Key::new(decoded.get(PUBLIC).unwrap().to_vec());
+            let iv = decoded.get(IV).unwrap();
 
-            let secret = std::str::from_utf8(&secret).unwrap().to_string();
+            let aes_key = Key::generate_aes_key(&self.private_key, &public_key);
+            encrypted = crypto::decrypt(secret, &aes_key, &iv).to_vec();
 
-            exchange_out.insert(SECRET, secret);
+            map.insert(SECRET, &encrypted);
         }
-        exchange_out.insert(PUBLIC, BASE64_STANDARD.encode(&self.public_key));
+        map.insert(PUBLIC, self.public_key.as_ref());
 
-        Ok(exchange_out)
+        encode(&map)
     }
 
-    pub fn secret_exchange_send(
-        &self,
-        secret: &str,
-        exchange: HashMap<&str, String>,
-    ) -> Result<HashMap<&str, String>, base64::DecodeError> {
-        let public_key = BASE64_STANDARD.decode(exchange.get(PUBLIC).unwrap())?;
-        let public_key = Key::new(public_key);
+    // Send Secret and perform encryption
+    pub fn secret_exchange_send(&self, secret: &str, exchange: String) -> String {
+        let decoded = decode(&exchange).unwrap();
```

Comment:
It's better not to take an owned value: `exchange: &str`?

Reply:
done, thanks :)